packages/angular/build/src/tools/vite/middlewares/assets-middleware.ts

```diff
@@ -69,13 +73,52 @@ export function createAngularAssetsMiddleware(
     if (extension !== '.js' && extension !== '.html') {
       const outputFile = outputFiles.get(pathname);
       if (outputFile?.servable) {
+        const data = outputFile.contents;
+        if (extension === '.css') {
+          // Inject component ID for view encapsulation if requested
+          const componentId = new URL(req.url, 'http://localhost').searchParams.get('ngcomp');
+          if (componentId !== null) {
+            // Record the component style usage for HMR updates
+            const usedIds = usedComponentStyles.get(pathname);
+            if (usedIds === undefined) {
+              usedComponentStyles.set(pathname, [componentId]);
+            } else {
+              usedIds.push(componentId);
+            }
+            // Shim the stylesheet if a component ID is provided
+            if (componentId.length > 0) {
+              // Validate component ID
+              // TODO: Determine if this regex is too strict for the APP_ID segment
+              if (/[.-A-Za-z0-9]+-c\d{9}$/.test(componentId)) {
```

Comment:
Technically for the app id portion it can be anything but I think this should be good enough. Consider adding `_` though to the regexp as that is differently used.